envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/JwtFilterFactory.kt

```diff
@@ -74,54 +76,41 @@ class JwtFilterFactory(
         .build()
 
-    private fun createRules(endpoints: List<IncomingEndpoint>): List<RequirementRule> {
-        return endpoints.mapNotNull(this::createRuleForEndpoint)
+    private fun createRules(endpoints: List<IncomingEndpoint>): Set<RequirementRule> {
+        return endpoints.mapNotNull(this::createRuleForEndpoint).toSet()
     }
 
     private fun createRuleForEndpoint(endpoint: IncomingEndpoint): RequirementRule? {
-        return if (endpoint.oauth != null) {
+        val providers = mutableSetOf<String>()
+
+        if (endpoint.oauth != null) {
+            providers.add(endpoint.oauth.provider)
+        }
+
+        providers.addAll(endpoint.clients.filter { it.selector in selectorToOAuthProvider.keys }
+            .mapNotNull { selectorToOAuthProvider[it.selector] })
+
+        return if (providers.isNotEmpty()) {
             RequirementRule.newBuilder().setMatch(
                 RouteMatch.newBuilder().setPrefix(endpoint.path)
             ).setRequires(
-                providersPolicyToJwtRequirements[endpoint.oauth.provider to endpoint.oauth.policy]
+                createJwtRequirement(providers)
             ).build()
         } else {
             null
         }
     }
 
-    private fun createProviderPolicyToJwtRequirements(): Map<Pair<String, OAuth.Policy>, JwtRequirement> {
-        return properties.providers.keys.flatMap { providerName ->
-            OAuth.Policy.values()
-                .map { policy -> Pair(providerName, policy) to createJwtRequirement(providerName, policy) }
-        }.associateBy({ it.first }, { it.second })
-    }
-
-    private fun createJwtRequirement(provider: String, policy: OAuth.Policy): JwtRequirement {
-        return when (policy) {
-            OAuth.Policy.ALLOW_MISSING -> {
-                JwtRequirement.newBuilder().setRequiresAny(
-                    JwtRequirementOrList.newBuilder().addAllRequirements(
-                        listOf(
-                            JwtRequirement.newBuilder().setProviderName(provider).build(),
-                            JwtRequirement.newBuilder().setAllowMissing(Empty.getDefaultInstance()).build()
-                        )
-                    )
-                ).build()
-            }
-            OAuth.Policy.ALLOW_MISSING_OR_FAILED -> {
-                JwtRequirement.newBuilder().setRequiresAny(
-                    JwtRequirementOrList.newBuilder().addAllRequirements(
-                        listOf(
-                            JwtRequirement.newBuilder().setProviderName(provider).build(),
-                            JwtRequirement.newBuilder().setAllowMissingOrFailed(Empty.getDefaultInstance()).build()
-                        )
-                    )
-                ).build()
-            }
-            OAuth.Policy.STRICT -> {
-                JwtRequirement.newBuilder().setProviderName(provider).build()
-            }
-        }
+    private fun createJwtRequirement(providers: Set<String>): JwtRequirement {
```

Comment:
Maybe we can create a map of providers and later instead of building them just retrieve them from map? It looks like you can only have providers defined in EC configuration.

Reply:
I think we can't create a map of providers bc we don't know how many providers it will include (possibly many when clients with selectors are present). I've created map for `JwtRequirements` inside `JwtRequirementOrList`